pyo3-macros-backend/src/pymethod.rs

```diff
@@ -452,55 +454,55 @@ const __FLOAT__: SlotDef = SlotDef::new("Py_nb_float", "unaryfunc");
 const __BOOL__: SlotDef = SlotDef::new("Py_nb_bool", "inquiry").ret_ty(Ty::Int);
 
 const __IADD__: SlotDef = SlotDef::new("Py_nb_inplace_add", "binaryfunc")
-    .arguments(&[Ty::ObjectOrNotImplemented])
+    .arguments(&[Ty::Object])
     .extract_error_mode(ExtractErrorMode::NotImplemented)
     .return_self();
```

Comment:
@birkenfeld: in answer to https://github.com/PyO3/pyo3/pull/1945#issuecomment-955548364

These definitions state what the generated function signature will be and place constraints on what is allowed in the pymethod. For example this `__iadd__` definiion will require there to be only one argument, of any type, and the return type to be `()` or `PyResult<()>`.

If the user deviates from these constraints there should be a compile error, however the error messages can probably be improved.